### ISSUE TITLE
[2.0.0-rc0] Using dataset makes layers.GRU not use cuDNN backend in graph mode

### ISSUE THREAD

**Issue by OlavHN** [tensorflow/tensorflow#32087]
I'm experiencing some weird behaviour when compiling a graph containing a GRU cuDNN operation and a dataset using `2.0.0-rc0`.

In eager execution, both executing the GRU and executing the GRU in a dataset loop performs similarly, doing 10 iterations in ~300ms.

When compiling to graph, using `tf.function` the time jumps to 3s for repeatedly exeuting the GRU, and to 30s when wrapped in a dataset iterator. Even when not using any data provided by the dataset (see example).

Looking at the trace it seems like the second case (graph + tf.data iterator: 100x slowdown) is due to it no longer using the cuDNN implementation, as the fused_matmul, sigmoid, etc operations are visible in the trace. In the first case (graph, no tf.data iterator: 10x slowdown) I'm not sure what causes the performance regression.

Attached is the profile traces generated by the script below: 

[gru_cudnn_dataset_graph_profiles.zip](https://github.com/tensorflow/tensorflow/files/3556137/gru_cudnn_dataset_graph_profiles.zip)

```
import tensorflow as tf
from tensorflow.python.eager import profiler

print(tf.version.VERSION, tf.version.GIT_VERSION)

ds = tf.data.experimental.Counter()

rnn = tf.keras.layers.GRU(1024)
inp = tf.random.uniform((128,100,512))

@tf.function
def f():
    start = tf.timestamp()
    for i in tf.range(10):
        rnn(inp)
        tf.print(tf.timestamp() - start)
        start = tf.timestamp()

with profiler.Profiler('test/graph'):
    print('graph')
    f()

def f():
    start = tf.timestamp()
    for i in tf.range(10):
        rnn(inp)
        tf.print(tf.timestamp() - start)
        start = tf.timestamp()

with profiler.Profiler('test/eager'):
    print('eager')
    f()

def f():
    start = tf.timestamp()
    for i in ds.take(10):
        rnn(inp)
        tf.print(tf.timestamp() - start)
        start = tf.timestamp()

with profiler.Profiler('test/dataset_eager'):
    print('dataset eager')
    f()

@tf.function
def f():
    start = tf.timestamp()
    for i in ds.take(10):
        rnn(inp)
        tf.print(tf.timestamp() - start)
        start = tf.timestamp()

with profiler.Profiler('test/dataset_graph'):
    print('dataset graph')
    f()
```
```
2.0.0-rc0 v2.0.0-beta1-5101-gc75bb66
graph
0.0048191547393798828
0.35352301597595215
0.36730408668518066
0.27707195281982422
0.36916518211364746
0.370150089263916
0.31262898445129395
0.28979992866516113
0.300137996673584
0.33440613746643066
eager
0.010959863662719727
0.011440038681030273
0.010359048843383789
0.011663913726806641
0.010751008987426758
0.00952005386352539
0.011648893356323242
0.010584115982055664
0.010535001754760742
0.010977983474731445
dataset eager
0.016921043395996094
0.01053309440612793
0.0098860263824462891
0.010875940322875977
0.0098860263824462891
0.0096909999847412109
0.0097639560699462891
0.00999903678894043
0.0095009803771972656
0.0097169876098632812
dataset graph
0.014286041259765625
3.5808620452880859
3.4670181274414062
3.5227558612823486
3.4991629123687744
3.4808690547943115
3.4846518039703369
3.5459158420562744
3.4938459396362305
3.4940009117126465
```

**Comment by kkimdev** [comment]
This is because for some reason @tf.function puts the work on CPU, not GPU.  If you change the last function to the following, you will observe a similar performance.

```python
@tf.function
def f():
    start = tf.timestamp()
    for i in ds.take(10):
        with tf.device('/device:GPU:0'):
            rnn(inp)
        tf.print(tf.timestamp() - start)
        start = tf.timestamp()
```

This issue will be tracked by https://github.com/tensorflow/tensorflow/issues/32138
